File: LeetCode/Study Plan/Practice Problems/Medium/Word Search/Go/word_search.go
package main

import "fmt"

var dir = [][]int{{0, 1}, {0, -1}, {1, 0}, {-1, 0}}

func search(grid [][]byte, word string) bool {
	n, m := len(grid), len(grid[0])
	visited := make([][]bool, n)
	for i := 0; i < n; i++ {
		visited[i] = make([]bool, m)
	}

	for i := 0; i < n; i++ {
		for j := 0; j < m; j++ {
			if word[0] == grid[i][j] {
				if dfs(i, j, grid, &visited, word, 0) {
					return true
				}
			}
		}
	}
	return false
}

func dfs(i, j int, grid [][]byte, visited *[][]bool, word string, k int) bool {
	if k == len(word) {
		return true
	}
	if i >= len(grid) || j >= len(grid[0]) || i < 0 || j < 0 || (*visited)[i][j] || k > len(word) {
		return false
	}
	if word[k] != grid[i][j] {
		return false
	}
	(*visited)[i][j] = true
	res := false
	for p := 0; p < 4; p++ {
		res = res || dfs(i+dir[p][0], j+dir[p][1], grid, visited, word, k+1)
	}
	(*visited)[i][j] = false
	return res
}

func main() {
	fmt.Println(search([][]byte{
		{'A', 'B', 'C', 'E'},
		{'S', 'F', 'C', 'S'},
		{'A', 'D', 'E', 'E'},
	}, "SEE"))
}

File: LeetCode/Study Plan/Practice Problems/Medium/Word Search/Java/wordSearch.java
class Solution {
    int n, m;
    int[][] dir = {{0, 1}, {0, -1}, {1, 0}, {-1, 0}};
    public boolean exist(char[][] board, String word) {
        n = board.length;
        m = board[0].length;
        
        if (n == 1 && m == 1 && word.length() == 1) return word.charAt(0) == board[0][0];

        boolean[][] visited = new boolean[n][m];
        for (int i=0; i<n; i++) {
            for (int j=0; j<m; j++) {
                if (board[i][j] == word.charAt(0)) {
                    if (wordSearch(i, j, board, word, visited)) return true;
                }
            }
        }
        return false;
    }

    private boolean wordSearch(int i, int j, char[][] board, String word, boolean[][] visited) {
        if (word.length() == 0) return true;
        
        if (word.charAt(0) != board[i][j])
            return false;
    
        if (visited[i][j]) return false;

        visited[i][j] = true;
        boolean res = false;
        for (int k=0; k<4; k++) {
            int x = i+dir[k][0];
            int y = j+dir[k][1];
            if (isSafe(x, y)) {
                res = res || wordSearch(x, y,board, word.substring(1), visited);
            }
        }
        visited[i][j] = false;
        return res;
    }

    private boolean isSafe(int i, int j) {
        return i>=0 && j>=0 && i<n && j<m;
    }
}
